From 831698091b0d4cbd7aa88522bb0fa5af3315cfe9 Mon Sep 17 00:00:00 2001
From: Mihails Strasuns <mihails.strasuns.contractor@sociomantic.com>
Date: Tue, 9 Feb 2016 22:24:24 +0200
Subject: [PATCH 03/13] Add message() to object.Throwable

---
 src/object.d | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/object.d b/src/object.d
index f0eb55a..e9ae3c6 100644
--- a/src/object.d
+++ b/src/object.d
@@ -1623,6 +1623,19 @@ class Throwable : Object
     string      msg;    /// A message describing the error.
 
     /**
+     * Get the message describing the error.
+     * Base behavior is to return the `Throwable.msg` field.
+     * Override to return some other error message.
+     *
+     * Returns:
+     *  message
+     */
+    const(char)[] message() const
+    {
+        return msg;
+    }
+
+    /**
      * The _file name and line number of the D source code corresponding with
      * where the error was thrown from.
      */
@@ -1689,6 +1702,8 @@ class Throwable : Object
         sink("@"); sink(file);
         sink("("); sink(sizeToTempString(line, tmpBuff, 10)); sink(")");
 
+        auto msg = this.message();
+
         if (msg.length)
         {
             sink(": "); sink(msg);
-- 
2.7.4

